src/backend/utils/misc/queryenvironment.c

```diff
@@ -978,8 +1095,279 @@ void ENRDropEntry(Oid id)
 
 	oldcxt = MemoryContextSwitchTo(currentQueryEnv->memctx);
 	currentQueryEnv->namedRelList = list_delete(currentQueryEnv->namedRelList, enr);
-	pfree(enr->md.name);
-	pfree(enr);
+
+	/* If we are dropping a committed ENR, wait until COMMIT to free it. */
+	if (temp_table_xact_support && enr->md.is_bbf_temp_table)
+	{
+		ListCell *lc;
+		enr->md.dropped_subid = GetCurrentSubTransactionId();
+		foreach(lc, currentQueryEnv->dropped_namedRelList)
+		{
+			EphemeralNamedRelation tmp_enr = (EphemeralNamedRelation) (lc);
+
+			if (strcmp(enr->md.name, tmp_enr->md.name) == 0 && tmp_enr->md.dropped_subid == enr->md.dropped_subid)
```

Comment:
given that we are handling this during rollback itself, we can skip this optimisation in favour of simplicity